### PR TITLE
json_writer: new routines to create JSON data

### PR DESCRIPTION
**Pull request by jeffhostetler**
Add "struct json_writer" and a series of jw_ routines to compose JSON
data into a string buffer.  The resulting string may then be printed by
commands wanting to support a JSON-like output format.

The json_writer is limited to correctly formatting structured data for
output.  It does not attempt to build an object model of the JSON data.

We say "JSON-like" because we do not enforce the Unicode (usually UTF-8)
requirement on string fields.  Internally, Git does not necessarily have
Unicode/UTF-8 data for most fields, so it is currently unclear the best
way to enforce that requirement.  For example, on Linux pathnames can
contain arbitrary 8-bit character data, so a command like "status" would
not know how to encode the reported pathnames.  We may want to revisit
this (or double encode such strings) in the future.

Helped-by: Eric Sunshine <sunshine@sunshineco.com>
Helped-by: René Scharfe <l.s.r@web.de>
Helped-by: Wink Saville <wink@saville.com>
Helped-by: Ramsay Jones <ramsay@ramsayjones.plus.com>
Signed-off-by: Jeff Hostetler <jeffhost@microsoft.com>

### CODE REVIEW

File: json-writer.c [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+		if (ch == '\n')
+			strbuf_addchars(sb, ' ', indent);
+	}
+}
+
+static void kill_indent(struct strbuf *sb,
+			const struct json_writer *jw)
+{
+	int k;
+	int eat_it = 0;
```
Comment (whoisj) [inline json-writer.c]: "eat_it" haha 😆  - maybe "skip" or "ignore"?

Comment (jeffhostetler) [reply json-writer.c]: I'll address it in the next version sent to the mailing list if I have to reroll it.  Thanks.

**Review by derrickstolee — APPROVED** [review]
Since this is merged into `next` I think this should be merged as-is. (Bonus points if they are the exact same commits as are in `jh/json-writer`.)

**Comment by jeffhostetler** [comment]
The json-writer code is in upstream master as of 2.19, so we don't need this PR.

**Pull request closed without merging**